src/nrniv/nrnbbcore_write.cpp

```diff
@@ -344,6 +351,30 @@ static void part2_clean() {
     delete artdata2index_;
     artdata2index_ = NULL;
   }
+
+  // clean up the art Memb_list of CellGroup[].mlwithart
+  // But if multithread and direct transfer mode defer deletion of
+  // data for artificial cells.
+  if (corenrn_direct && nrn_nthread > 0) {
+    deferred_type2artdata.resize(nrn_nthread);
+  }
```

Comment:
I improved the comment to:
```
  // clean up the art Memb_list of CellGroup[].mlwithart
  // But if multithread and direct transfer mode, defer deletion of
  // data for artificial cells, so that the artificial cell ml->data
  // can be used when nrnthreads_type_return is called.
```
I think it also helps to see the comment in nrnthreads_type_return, which gives the justification for all this.
```
     // The single thread case is easy
...
        // mk_tml_with_art() created a cgs[id].mlwithart which appended
        // artificial cells to the end. Turns out that
        // cellgroups_[tid].type2ml[type]
        // is the Memb_list we need. Sadly, by the time we get here, cellgroups_
        // has already been deleted.  So we defer deletion of the necessary
        // cellgroups_ portion (deleting it on return from nrncore_run).
```